New page about Kotlin daemon (#5044)
JetBrains/kotlin-web-site: docs/topics/kotlin-daemon.md

[//]: # (title: Kotlin daemon)

The Kotlin daemon is a background process that build systems can use to improve Kotlin compilation performance.
The Kotlin daemon speeds up build times by keeping the compiler and its environment ready to compile. This approach 
avoids the overhead of starting a new Java Virtual Machine (JVM) instance and reinitializing the compiler for every 
compilation, leading to reduced build times for incremental builds or frequent small changes.

Some build systems, like [Gradle](https://docs.gradle.org/current/userguide/gradle_daemon.html) and [Maven](https://maven.apache.org/tools/mvnd.html),
have their own daemons that help reduce startup costs. However, they miss the added benefit that the Kotlin daemon
completely isolates the build system process from the compiler. This separation is useful in dynamic environments where
system settings might change at runtime.

Although there's no official user-facing interface for the daemon, it can be used by build systems or through the [build tools API](build-tools-api.md).

## Kotlin daemon configuration

There are ways to configure some settings for the Kotlin daemon for Gradle or Maven.

### Memory management

> By default, the Kotlin daemon tries to inherit the heap size (`-Xmx`) of the launching JVM process.

Reviewer: ```suggestion
The Kotlin daemon is a separate process that has its own memory space, isolated from the client.
By default, the Kotlin daemon tries to inherit the heap size (`-Xmx`) of the launching JVM process.
```

To configure specific memory limits, like `-Xmx` and `-XX:MaxMetaspaceSize`, use the following property:

<tabs group="build-system">
<tab title="Gradle" group-key="gradle">

```kotlin
kotlin.daemon.jvmargs=-Xmx1500m
```

For more information, see [`kotlin.daemon.jvmargs` property](gradle-compilation-and-caches.md#kotlin-daemon-jvmargs-property).

</tab>
<tab title="Maven" group-key="maven">

```xml
<kotlin.compiler.daemon.jvmArgs>-Xmx1500m</kotlin.compiler.daemon.jvmArgs>
```

</tab>
</tabs>

### Lifetime

There are two common lifetime strategies for the Kotlin daemon:

> * **Short-lived daemon**: Used when the client (like a build system) is long-running. The daemon shuts down shortly after the

Reviewer: I'm not sure about the naming "short-lived" and "long-lived".
Just as an idea, we may call them "attached" and "detached", where attached has its lifetime attached to the client process and detached is based on timeouts and last usage

Author: Works for me! Added.

> * **Short-lived daemon**: Used when the client (like a build system) is long-running. The daemon shuts down shortly after the
> last compilation.

Reviewer: ```suggestion
* **Short-lived daemon**: Used when the client (like a build system) is long-running. The daemon shuts down shortly after the
client process shuts down or the daemon hasn't been used for a while.
```

* **Long-lived daemon**: Used when the client is short-lived (like command-line compilation). The daemon stays alive longer
to await potential follow-up requests.

To configure the lifetime strategy, you can use the following options:

> | Option | Description                                                                                        | Default value |
> |--------|----------------------------------------------------------------------------------------------------|---------------|
> | `autoshutdownIdleSeconds` | How long the daemon should stay alive after the last compilation when a client is still connected. | 2 hours |
> | `autoshutdownUnusedSeconds` | How long the daemon waits to shut down after all clients disconnect.                               | 1 minute |
> | `shutdownDelayMilliseconds` | How long a newly started daemon waits for a first client before shutting down if unused.           | 1 second |

Reviewer: ```suggestion
| Option                      | Description                                                                                        | Default value |
|-----------------------------|----------------------------------------------------------------------------------------------------|---------------|
| `autoshutdownIdleSeconds`   | How long the daemon should stay alive after the last compilation when a client is still connected. | 2 hours       |
| `autoshutdownUnusedSeconds` | How long a newly started daemon waits for a first client before shutting down if unused.           | 1 minute      |
| `shutdownDelayMilliseconds` | How long the daemon waits to shut down after all clients disconnect.                               | 1 second      |
```

> To configure a short-lived daemon lifetime, set `shutdownDelayMilliseconds` to a **high** value.

Reviewer: So in the terminology of "attached" and "detached" this should be "detached"

<tabs group="build-system">
<tab title="Gradle" group-key="gradle">

Add the following to your `gradle.properties` file:

```none
org.gradle.jvmargs=-Dkotlin.daemon.jvm.options=shutdownDelayMilliseconds=7200
```

</tab>
<tab title="Maven" group-key="maven">

Add the following property to your `pom.xml` file:

```xml
<kotlin.compiler.daemon.shutdownDelayMs>7200</kotlin.compiler.daemon.shutdownDelayMs>
```

</tab>
</tabs>

> To configure a long-lived daemon lifetime, set `autoshutdownIdleSeconds` to a **high** value and `shutdownDelayMilliseconds` to a **low** value.

Reviewer: and this should be "attached"

[//]: # (title: Kotlin daemon)

The Kotlin daemon is a background process that build systems can use to improve Kotlin compilation performance.
The Kotlin daemon speeds up build times by keeping the compiler and its environment ready to compile. This approach 
avoids the overhead of starting a new Java Virtual Machine (JVM) instance and reinitializing the compiler for every 
compilation, leading to reduced build times for incremental builds or frequent small changes.

Some build systems, like [Gradle](https://docs.gradle.org/current/userguide/gradle_daemon.html) and [Maven](https://maven.apache.org/tools/mvnd.html),

> have their own daemons that help reduce startup costs. However, they miss the added benefit that the Kotlin daemon

Author: You can use the build system daemon with the "in-process" compiler execution strategy or you can use the Kotlin daemon compiler execution strategy. The Kotlin daemon has the benefit of isolating the build system process from the compiler. I tried to adjust the text to make it clearer that it's one or the other. Does it read better to you?

completely isolates the build system process from the compiler. This separation is useful in dynamic environments where
system settings might change at runtime.

> Although there's no official user-facing interface for the daemon, it can be used by build systems or through the [build tools API](build-tools-api.md).

Reviewer: I'm only not sure about this wording.
It's expected that build system plugins may expose daemon functionality, but they also are expected to work through the Build Tools API

There are ways to configure some settings for the Kotlin daemon for Gradle or Maven.

### Memory management

The Kotlin daemon is a separate process that has its own memory space, isolated from the client.
By default, the Kotlin daemon tries to inherit the heap size (`-Xmx`) of the launching JVM process.

To configure specific memory limits, like `-Xmx` and `-XX:MaxMetaspaceSize`, use the following property:

<tabs group="build-system">
<tab title="Gradle" group-key="gradle">

```kotlin
kotlin.daemon.jvmargs=-Xmx1500m
```

For more information, see [`kotlin.daemon.jvmargs` property](gradle-compilation-and-caches.md#kotlin-daemon-jvmargs-property).

</tab>
<tab title="Maven" group-key="maven">

```xml
<kotlin.compiler.daemon.jvmArgs>-Xmx1500m</kotlin.compiler.daemon.jvmArgs>
```

</tab>
</tabs>

### Lifetime

There are two common lifetime strategies for the Kotlin daemon:

* **Attached daemon**: Used when the client (like a build system) is long-running. The daemon shuts down shortly after the
client process shuts down or the daemon hasn't been used for a while.
* **Detached daemon**: Used when the client is short-lived (like command-line compilation). The daemon stays alive longer

> to await potential follow-up requests.

Reviewer 2: After this, I expected "attached daemon / detached daemon" or "lifetime strategy" to be specific programmatic categories, which are to be used. But as I understand the user first needs to implement a strategy, then use it. So maybe it would be clearer like this:

```
There are two common strategies that you can implement for the Kotlin daemon:

* Attached daemon: Shut down the daemon shortly after <...>. Use when the client is long-running.
* Detached daemon: Keep the daemon alive longer <...>. Use when the client is short-lived.
```

To configure the lifetime strategy, you can use the following options:

| Option                      | Description                                                                                        | Default value |
|-----------------------------|----------------------------------------------------------------------------------------------------|---------------|
| `autoshutdownIdleSeconds`   | How long the daemon should stay alive after the last compilation when a client is still connected. | 2 hours       |
| `autoshutdownUnusedSeconds` | How long a newly started daemon waits for a first client before shutting down if unused.           | 1 minute      |
| `shutdownDelayMilliseconds` | How long the daemon waits to shut down after all clients disconnect.                               | 1 second      |

> To configure an attached daemon lifetime, set `autoshutdownIdleSeconds` to a **high** value and `shutdownDelayMilliseconds` to a **low** value.

Reviewer 2: Either lifetime settings define whether the daemon is attached or detached, or I missed the point somewhere around here 😅
